src/feat/pitch-functions.h

```diff
@@ -198,12 +198,16 @@ struct PitchExtractionOptions {
   }
   /// Returns the window-size in samples, after resampling.  This is the
   /// "basic window size", not the full window size after extending by max-lag.
+  // B.P. fixed a bug here. Even though the result of the operation is supposed 
+  // to be an integer, you still need rounding before casting :-)
+  // NB: 0.001 is impossible to represent exactly on a float or double, so for
+  // this to work without rounding, you would need to divide by 1000 instead.
   int32 NccfWindowSize() const {
-    return static_cast<int32>(resample_freq * 0.001 * frame_length_ms);
+      return static_cast<int32>(round(resample_freq * 0.001f * frame_length_ms));
```

Comment:
This one is quite obscur that probably deserves a detailed explanation. Hope you have the time :-)

For example, consider frame_length_ms=0.9, resample_freq=10000. I would expect the formula above to give me a WindowSize of "9", but it actually gives "8" on my Intel box, and probably something different on ARM CPUs. In practice, you will get potentially unpredictable results whenever one of the operand is not perfectly representable on a floating point (e.g. anything in decimal such as 0.1 or 0.001).

This has for example some impacts on the number of samples generated, that will not match what other featbin programs generate. The WindowShift formula is potentially even more problematic. And yes, it is a bug I ran into, and it was a pain to figure out where it came from.

The issue here is that rounding down with float in general is dangerous, as you can quickly get into some unexpected situations such as the one above. This make things hard to reproduce externally, as many programming languages with the same formula would actually give you 9 (as you would expect).

My reasoning is that I would expect it to give me the "right" integer when the formula is supposed to give you an "exact" integer value, but as it is written, the result heavily depends on the internal floating point representation of frame_length_ms or resample_freq. The result of that formula as it is written is "hardware floating-point" implementation dependent in many cases, which is *never* something desirable, especially if you want kaldi to give reproducible results.

Now, the question is how do you fix it; dividing by 1000 rather than multiplying by 0.001 actually fixes pretty much any situation that will occur on Intel CPUs, but that makes the fix rather obscure. Furthermore, rounding to the closest is safer and probably closer to what most people would expect anyway:
 - it fixes the border cases above
 - it doesn't change the behaviour for "normal" cases
 - it is probably faster that doing a floating point division
 - it is _probably_ - but that could be discussed-  theoretically more useful than rounding down.

And sorry about the indentation! My editor was not properly configured to generate 2 spaces indentations.